_posts/2017-08-16-json-server.md

```diff
@@ -0,0 +1,655 @@
+---
+layout: post
+title: JSON Server
+excerpt: Accelerate your prototyping process by mocking your APIs using JSON Server
+authors:
+    - kelfarsaoui
+permalink: /en/json-server/
+categories:
+    - JSON
+    - API
+    - NodeJS
+tags:
+    - JSON
+    - API
+    - NodeJS
+cover: /assets/2017-08-16-json-server/cover.jpg
+---
+
+Hello, Today I’m going to talk about [json-server](https://github.com/typicode/json-server), what is it? why use it? and especially how to use it?
+
+`json-server` is a cool `npm` module, that provides an [Express](https://github.com/expressjs/express) server that serves a JSON API.
+
+## Why use it?
+
+Let's say you are starting to work on your awesome (Javascript, PHP, IOS, Android, … whatever) app, and you want to consume remote data of some, not finished yet, API. The first step you take consists in setting up some mocks, either hard coded in some constant (please, don't do this), or using a static JSON file, which will be a pain in the neck. Why? You're a good developer who likes to do the right thing. You want to test your app against data updates. You want to perform HTTP requests (like `GET`, `POST` …etc.), and you want to persist your updates. Unfortunately, you can't do all of this using a static file, you need to look for a way to make it dynamic. So, unless your pal who's developing the API has finished their work, you're gonna need some serious help.
+
+Here comes `json-server`, it allows you to mock the API, and provide dynamic access to data. By dynamic, I mean that we can read, add, update, and delete data (`GET`, `POST`, `PUT`, `PATCH`, `DELETE`).
+
+It provides common URL use cases like:
+- [Routes](#routes) (`/articles/1`)
+- [Filters](#filters) (`/articles/1/comments?author.username=rpierlot`)
+- [Pagination](#pagination) (`/articles?_page=2&_limit=10`)
+- [Full text search](#full-text-search) (`/articles?q=graphql`)
+- [Relationships](#relationships) (`/articles?_embed=comments`)
+
+and other miscellaneous stuff like:
+- `CORS` & `JSONP`
+- Ability to access remote schemas
+- [Randomly generated data](#random-data)
+- [Custom routes](#custom-routes)
+- [Middlewares](#middlewares)
+- [Ability to use `json-server` as a module in your NodeJS server](#nodejs-module)
+
+Okay, let's get to it.
+
+## How to use it?
+
+This can take less than 5min!
+
+### Requirements
+
+- NodeJS & `npm`
+- An API consumer (Your code, `curl`, postman, or simply your browser)
+
+### Installation
+
+It's pretty straightforward to set up:
+
+```bash
+$ npm install -g json-server
+
+# Or with yarn
+$ yarn global add json-server
+
+# Then create a folder in which we will put our db.json file:
+$ mkdir blog && cd $_
+
+# create our schema file
+$ touch db.json
+```
+
+To fill it in, we can either do it by hand or use a random json generator, my favorite is [json-generator](http://json-generator.com)
```

Comment:
To fill it in, we can either do it by hand or use a random json generator (my favorite is [json-generator](http://json-generator.com) )